### ISSUE TITLE
About `Dens3R`'s Visual Quality Improvement

### ISSUE THREAD

**Issue by DengKaiCQ** [G-1nOnly/Dens3R#4]
Hi my friend,  

Thank you for your excellent work on `Dens3R`.

From the visuals, `Dens3R` demonstrates significantly finer depth precision compared to `VGGT` (from fig 18~20), though quantitative results in the table 3 suggest comparable performance between the two methods. Regarding this, the `Dens3R` attributes the improvement to its normal-based training strategy to my understanding. Intro and method parts shows this strategy is a key factor in enhancing the quality of depth predictions.

<div align="center">
<img width="65%" alt="Image" src="https://github.com/user-attachments/assets/9255bae0-f5cb-4eca-8d46-757cd62ae1c1" />

<img width="65%" alt="Image" src="https://github.com/user-attachments/assets/44257252-436f-4bf9-a22c-614d4f7d12c5" />
</div>

While `VGGT` does not explicitly use normal loss, its training objective includes a 1st-order derivative loss of depth (fig below). Since surface normals are fundamentally equivalent to the normalized gradient of depth, `VGGT` implicitly incorporates normal information during training. 

<div align="center">
<img width="45%" alt="Image" src="https://github.com/user-attachments/assets/87250316-6203-4bd1-9a71-84740c0ad4ce" />
</div>

<div align="center">
(Sec 3.4 in the paper of `VGGT`)
</div>

My question lies this part: is the superior depth refinement in `Dens3R` (fig 18~20) primarily attributable to the explicit normal loss? Or does it stem more fundamentally from the strategy briefly mentioned in eq. 9?

Additionally, I have a another question that what is the max num of imgs `Dens3R` can process in a single forward pass? Can it handle tens of imgs per forward pass like `Fast3R` or `VGGT`? If I understand this correctly, `Dens3R` processes two imgs per a single forward pass, with multi-view inference achieved through post-processing via imgs pair matches (Sec. 3.3).

Finally, thank you again for your impactful contribution. The geometric detail in `Dens3R`’s depth is a valuable training strategy reference.

**Comment by G-1nOnly** [comment]
Hi @DengKaiCQ, thank you for your interest in our work! In my opinion, Dens3R improves visual quality primarily through the normal-based training strategy and the coarse-to-fine training.

For the normal-based training strategy, we employ an explicit normal loss to supervise the view-space normals. We observed that this leads to more accurate and sharper predictions. In addition, we remove the confidence loss since we found that with this loss the model tends to avoid challenging regions such as reflective surfaces. Furthermore, we adopt the intrinsic-invariant pointmap (Eq. 9) to enforce one-to-one correspondences, which simplifies training because we do not need to regress across views as in other 3R-based methods. With this one-to-one property, the depth head can be further fine-tuned using both multi-view and monocular data, leading to improved results.

As for the coarse-to-fine training, we find that including high-resolution data (e.g., 1024×1024) is crucial for achieving high-quality depth and normal predictions. Not only the model better captures fine details, but it also enables higher-resolution inference, as mentioned in our paper. High-quality data is therefore essential for training models that produce high-fidelity visual predictions. This observation is also consistent with MoGe-v2, a concurrent work that advances monocular depth estimation. Additionally, because the ground-truth annotations in the evaluation dataset are not of high quality, the quantitative results across methods may appear comparable—this can also be seen in the results of MoGe-v2.

Thank you again for the thoughtful questions, and we hope our explanation addresses your concerns.
